src/main/java/com/google/sps/servlets/UploadServlet.java

```diff
@@ -0,0 +1,150 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+ 
+package com.google.sps.servlets;
+ 
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Collection;
+import javax.servlet.ServletException;
+import javax.servlet.annotation.MultipartConfig;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import javax.servlet.http.Part;
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.nio.file.Files;
+import com.google.cloud.storage.BlobId;
+import com.google.cloud.storage.BlobInfo;
+import com.google.cloud.storage.Storage;
+import com.google.cloud.storage.StorageOptions;
+import java.nio.file.Paths;
+
+/** Servlet for uploading files. */
+@WebServlet("/upload")
+@MultipartConfig
+public class UploadServlet extends HttpServlet {
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException, ServletException {
+    //  Get all the Dataset Entities
+    Query query = new Query("MetaData").addSort("timestamp", SortDirection.ASCENDING);
+
+    /* Get current user logged in to webapp */
+    UserService userService = UserServiceFactory.getUserService();
+
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    PreparedQuery results = datastore.prepare(query);
+
+    /*  Get metadata (for comparison and eventual storage) */
+    String userEmail = userService.getCurrentUser().getEmail();
+    String datasetName = request.getParameter("dataset-namer");
+    long timestamp = System.currentTimeMillis();
+
+    // Loop to see if that user already exists
+    boolean datasetPresent = false;
+    boolean userPresent = false;
+    boolean newDataset = true;  // Assume this is a new dataset being created
+    for (Entity entity : results.asIterable()) {
```

Comment:
How about factoring this part out as a separate method (e.g., IsDatasetPresent(..)) and also create a unit test?

Reply:
(Maybe you don't need this if you will further refactor this CL with a separate datastore class).

Reply:
As we discussed, I'll save this for a separate PR after this, including my own tests